internal/pkg/service/appsproxy/http/router.go

```diff
@@ -118,13 +140,15 @@ func (r *Router) CreateHandler() http.Handler {
 		}
 
 		// Recreate app handler if configuration changed.
-		handler, ok := r.appHandlers[appID]
-		if !ok || modified {
-			handler = r.createDataAppHandler(req.Context(), config)
-			r.appHandlers[appID] = handler
+		handler := r.appHandlers.GetOrInit(appID)
+		httpHandler := handler.getHTTPHandler()
+
+		if modified || httpHandler == nil {
+			httpHandler = r.createDataAppHandler(req.Context(), config)
+			handler.setHTTPHandler(httpHandler)
 		}
```

Comment:
`httpHandler = r.createDataAppHandler(req.Context(), config)` part is not protected by the lock and can run more times.

Maybe we can use something like following code?
```
httpHandler := handler.getOrInit(
    return r.createDataAppHandler(req.Context(), config)
)
```

I understand that the problem is `ctx` and `config`.

Reply:
You can implement this use-case specially. Or modify `SafeMap` so that the `init callback` is passed as an argument to the `GetOrInit` method.